### ISSUE TITLE
[no-restricted-imports] 2 pattern rules with different globs and `allowTypeImports` break validation

### ISSUE THREAD

**Issue by elado** [typescript-eslint/typescript-eslint#4895]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/<rule>": ["error", {
	    "patterns": [
	      {
	        "group": [
	          "import1/private/*"
	        ],
	        "message": "usage of import1 private modules not allowed.",
	        "allowTypeImports": true
	      },
	      {
	        "group": [
	          "import2/private/*"
	        ],
	        "message": "usage of import2 private modules not allowed.",
	        "allowTypeImports": true
	      }
	    ]
	  }]
  }
}
```

```TS
import type {foo} from 'import1/private/bar'; 
import type {foo} from 'import2/private/bar';
```

**Expected Result**

Both type imports should be valid.

**Actual Result**

Both imports fail lint.

**Additional Info**

If only one `allowTypeImports` pattern rule exists, it works. Once I enable two rules, both break.

A failing test case for [no-restricted-imports.test.ts](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/tests/rules/no-restricted-imports.test.ts):

```ts
    {
      code: `
        import type {foo} from 'import1/private/foo';
        import type {bar} from 'import2/private/bar';
      `,
      options: [
        {
          patterns: [
            {
              group: ['import1/private/*'],
              message: 'usage of import1 private modules not allowed.',
              allowTypeImports: true,
            },
            {
              group: ['import2/private/*'],
              message: 'usage of import2 private modules not allowed.',
              allowTypeImports: true,
            },
          ],
        },
      ],
    },
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.22.0` |
| `@typescript-eslint/parser`        | `5.22.0` |
| `TypeScript`                       | `X.Y.Z` |
| `ESLint`                           | `X.Y.Z` |
| `node`                             | `X.Y.Z` |
